Web3/Classes/Core/Contract/SolidityInvocation.swift

```diff
@@ -0,0 +1,243 @@
+        guard let data = encodeABI() else { return nil }
+        guard let to = handler.address else { return nil }
+        return EthereumTransaction(nonce: nonce, gasPrice: gasPrice, gas: gas, from: from, to: to, value: value ?? 0, data: data)
+    }
+    
+    public func send(nonce: EthereumQuantity? = nil, from: EthereumAddress, value: EthereumQuantity?, gas: EthereumQuantity, gasPrice: EthereumQuantity?, completion: @escaping (EthereumData?, Error?) -> Void) {
+        guard handler.address != nil else {
+            completion(nil, InvocationError.contractNotDeployed)
+            return
+        }
+        guard let transaction = createTransaction(nonce: nonce, from: from, value: value, gas: gas, gasPrice: gasPrice) else {
+            completion(nil, InvocationError.encodingError)
+            return
+        }
+        handler.send(transaction, completion: completion)
+    }
+}
+
+// MARK: - Non Payable Invocation
+
+/// An invocation that writes to the blockchain and cannot receive ETH. Should only use .send().
+public struct SolidityNonPayableInvocation: SolidityInvocation {
+    public let method: SolidityFunction
+    public let parameters: [SolidityWrappedValue]
+    
+    public let handler: SolidityFunctionHandler
+    
+    public init(method: SolidityFunction, parameters: [ABIConvertible], handler: SolidityFunctionHandler) {
+        self.method = method
+        self.parameters = zip(parameters, method.inputs).map { SolidityWrappedValue(value: $0, type: $1.type) }
+        self.handler = handler
+    }
+    
+    public func createTransaction(nonce: EthereumQuantity? = nil, from: EthereumAddress, value: EthereumQuantity?, gas: EthereumQuantity, gasPrice: EthereumQuantity?) -> EthereumTransaction? {
+        guard let data = encodeABI() else { return nil }
+        guard let to = handler.address else { return nil }
+        return EthereumTransaction(nonce: nonce, gasPrice: gasPrice, gas: gas, from: from, to: to, value: value ?? 0, data: data)
+    }
+    
+    public func send(nonce: EthereumQuantity? = nil, from: EthereumAddress, value: EthereumQuantity?, gas: EthereumQuantity, gasPrice: EthereumQuantity?, completion: @escaping (EthereumData?, Error?) -> Void) {
+        guard handler.address != nil else {
+            completion(nil, InvocationError.contractNotDeployed)
+            return
+        }
+        guard let transaction = createTransaction(nonce: nonce, from: from, value: value, gas: gas, gasPrice: gasPrice) else {
+            completion(nil, InvocationError.encodingError)
+            return
+        }
+        handler.send(transaction, completion: completion)
+    }
+}
+
+// MARK: - PromiseKit convenience
+
+public extension SolidityInvocation {
+    
+    // Default Implementations
+    
+    public func createCall() -> EthereumCall? {
```

Comment:
What are the default implementations for if they actually never do something useful?

Reply:
At least a few of them in this protocol

Reply:
This is sort of just a convenience for the other structs so they can focus their implementation on the methods they support. Pros and cons here, the biggest con being that if there was another implementer they may not realize they are missing a method.